python: add mesonpep517 as build system

FRidh opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

I would like to demonstrate mesonpep517 and thus choose to upstream it.

While it "works", I am not very happy with this yet. See e.g. the comments in the code.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

FRidh committed:
pythonPackages.intreehooks: init at 1.0

FRidh committed:
pythonPackages.mesonpep517: init at 0.1.9999994

FRidh commented on pkgs/development/python-modules/mesonpep517/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  propagatedNativeBuildInputs = [ meson ninja ];
+
+  # This should be propagated as long as meson comes with a hook.
```
I am starting to get an increasing dislike of packages *including* a setup hook when you do not want it. Maybe we should split it off, that is into a `meson` and `mesonHook`.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/trouble-building-passwordsafe-a-meson-python-gtk3-package/4652/6

FRidh commented on pkgs/development/python-modules/mesonpep517/default.nix:
```diff
@@ -0,0 +1,46 @@
+  meta = {
+    description = "Create pep517 compliant packages from the meson build system";
+    homepage = https://gitlab.com/thiblahute/mesonpep517;
+    license = lib.licenses.asl;
```
incorrect license

jtojnar commented:
We will need to patch this to support multiple outputs:

https://gitlab.com/thiblahute/mesonpep517/blob/d3ca51c80a7960b71b2f8852c235d9269cf2e038/mesonpep517/buildapi.py#L44-51

FRidh commented on pkgs/development/python-modules/mesonpep517/default.nix:
```diff
@@ -0,0 +1,46 @@
+, pythonOlder
+}:
+
+# TODO: offer meson as a Python package so we have dist-info folder.
```
It's silly, actually. We need it is as a Python package if packages start adding it as a build system requirement. However, it should not be needed considering meson is only an *application*; they do not actually import it.

FRidh committed:
pythonPackages.meson: init

Add meson as Python package so it can easily be used by mesonpep517 for
PEP 517 builds.

FRidh commented:
@grahamcofborg build python3.pkgs.mesonpep517
